readthedocs/projects/tasks/builds.py

```diff
@@ -0,0 +1,1020 @@
+                version_type=self.version.type,
+                build_pk=self.build['id'],
+                commit=self.commit,
+                status=BUILD_STATUS_FAILURE,
+            )
+
+        # Update build object
+        self.build['success'] = False
+
+    def on_success(self, retval, task_id, args, kwargs):
+        html = self.outcomes['html']
+        search = self.outcomes['search']
+        localmedia = self.outcomes['localmedia']
+        pdf = self.outcomes['pdf']
+        epub = self.outcomes['epub']
+
+        # Store build artifacts to storage (local or cloud storage)
+        self.store_build_artifacts(
+            self.build_env,
+            html=html,
+            search=search,
+            localmedia=localmedia,
+            pdf=pdf,
+            epub=epub,
+        )
+
+        # NOTE: we are updating the db version instance *only* when
+        # HTML are built successfully.
+        if html:
+            try:
+                api_v2.version(self.version.pk).patch({
+                    'built': True,
+                    'documentation_type': self.get_final_doctype(),
+                    'has_pdf': pdf,
+                    'has_epub': epub,
+                    'has_htmlzip': localmedia,
+                })
+            except HttpClientError:
+                # NOTE: I think we should fail the build if we cannot update
+                # the version at this point. Otherwise, we will have inconsistent data
+                log.exception(
+                    'Updating version failed, skipping file sync.',
+                )
+
+        # Index search data
+        fileify.delay(
+            version_pk=self.version.pk,
+            commit=self.build['commit'],
+            build=self.build['id'],
+            search_ranking=self.config.search.ranking,
+            search_ignore=self.config.search.ignore,
+        )
+
+        if not self.project.has_valid_clone:
+            self.set_valid_clone()
+
+        self.send_notifications(
+            self.version.pk,
+            self.build['id'],
+            event=WebHookEvent.BUILD_PASSED,
+        )
+
+        if self.commit:
+            send_external_build_status(
+                version_type=self.version.type,
+                build_pk=self.build['id'],
+                commit=self.commit,
+                status=BUILD_STATUS_SUCCESS,
+            )
+
+        # Update build object
+        self.build['success'] = True
+
+        # NOTE: we can probably remove the ``exit_code`` at build level, I
+        # don't think we are using it
+        # self.build['exit_code'] = max([
+        #     cmd.exit_code for cmd in self.build_env.commands
+        # ])
+
+        # NOTE: I don't think we are using these fields either
+        # self.build['setup'] = self.build['setup_error'] = ''
+        # self.build['output'] = self.build['error'] = ''
+
+    def on_retry(self, exc, task_id, args, kwargs, einfo):
+        log.warning('Retrying this task.')
```

Comment:
Just curious, I've seen this a few times. Is information on the task that is executing bound somewhere already? That is, will log lines include some metadata on the task and task_id in our logs?

Reply:
Yes! This is the magic of `log.bind` from `structlog` (see https://www.structlog.org/en/stable/getting-started.html#building-ctx)

We are calling `log.bind` on `before_start` with all the useful data for this particular build: https://github.com/readthedocs/readthedocs.org/blob/d5a8ffb0e458e447772bc7d6008e5a39893f3af7/readthedocs/projects/tasks/builds.py#L329-L336

The following `log.info` will include all those keys + the message we are logging.

Besides, there are some "extra Celery metadata" that is added by the Django app `django-structlog`. For example, `task_id` (see https://github.com/jrobichaud/django-structlog/blob/master/django_structlog/celery/receivers.py)

Reply:
:magic_wand: